app/src/test/java/br/com/catbag/gifreduxsample/reducers/middlewares/RestMiddlewareTest.java

```diff
@@ -0,0 +1,213 @@
+import static br.com.catbag.gifreduxsample.actions.GifListActionCreator.GIF_LIST_UPDATED;
+import static br.com.catbag.gifreduxsample.asyncs.data.DataManager.GifListLoadListener;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static shared.TestHelper.buildGif;
+
+// Roboeletric still not supports API 24 stuffs
+@Config(sdk = 23, constants=BuildConfig.class)
+@RunWith(RobolectricTestRunner.class)
+public class RestMiddlewareTest extends ReduxBaseTest {
+
+    public static final boolean EXPECTED_HAS_MORE = false;
+    public static final List EXPECTED_LIST = new ArrayList<>();
+
+    private Fluxxan<AppState> mFluxxan;
+    private CountDownLatch mSignal;
+    private Action mLastAction;
+
+    private Middleware mInterceptMiddleware = new BaseMiddleware() {
+        @Override
+        public void intercept(Object o, Action action) throws Exception {
+            mLastAction = action;
+            mSignal.countDown();
+        }
+    };
+
+    public RestMiddlewareTest() {
+        AppState state = ImmutableAppState.builder().build();
+        mFluxxan = new Fluxxan<>(state);
+        mHelper = new TestHelper(mFluxxan);
+        mFluxxan.start();
+    }
+
+    @Before
+    public void setup() {
+        mSignal = new CountDownLatch(1);
+        mFluxxan.getDispatcher().registerMiddleware(mInterceptMiddleware);
+    }
+
+    @After
+    @Override
+    public void cleanup() {
+        super.cleanup();
+        mFluxxan.stop();
+        mFluxxan.getDispatcher().unregisterMiddleware(mInterceptMiddleware.getClass());
+    }
+
+    @Test
+    public void whenInterceptListFetching() throws Exception {
+        RestMiddleware middleware = new RestMiddleware(RuntimeEnvironment.application,
+                mockDataManager(), any(FileDownloader.class));
+        middleware.setDispatcher(mFluxxan.getDispatcher());
+        middleware.intercept(mFluxxan.getState(), new Action(GIF_LIST_FETCHING));
+        mSignal.await();
+
+        assertEquals(GIF_LIST_UPDATED, mLastAction.Type);
+        assertEquals(mLastAction.Payload.getClass(), HashMap.class);
+        Map params = (Map) mLastAction.Payload;
+        assertEquals(params.get(PayloadParams.PARAM_HAS_MORE), EXPECTED_HAS_MORE);
+        assertEquals(((List)params.get(PayloadParams.PARAM_GIFS)).size(), EXPECTED_LIST.size());
+    }
+
+    @Test
+    public void whenInterceptListFetchingAndDontHasMoreGifs() throws Exception {
+        AppState state = ImmutableAppState.builder().hasMoreGifs(false).build();
+        DataManager dataManager = mockDataManager();
+        RestMiddleware middleware = new RestMiddleware(RuntimeEnvironment.application,
```

Comment:
If you want...import `RuntimeEnviroment` statically and put `mockDataManager();` directly on constructor....